festim/hydrogen_transport_problem.py

```diff
@@ -123,11 +162,66 @@ def initialise(self):
         self.t = fem.Constant(self.mesh.mesh, 0.0)
         self.dt = self.settings.stepsize.get_dt(self.mesh.mesh)
 
+        self.define_temperature()
         self.define_boundary_conditions()
         self.create_formulation()
         self.create_solver()
         self.defing_export_writers()
 
+    def define_temperature(self):
+        """Sets the value of temperature_fenics_value. The type depends on
+        self.temperature. If self.temperature is a function on t only, create
+        a fem.Constant. Else, create an dolfinx.fem.Expression (stored in
+        self.temperature_expr) to be updated, a dolfinx.fem.Function object
+        is created from the Expression (stored in self.temperature_fenics_value)
+        """
+        # check if temperature is None
+        if self.temperature is None:
+            raise TypeError("Temperature needs to be defined")
+
+        # if temperature is a float or int, create a fem.Constant
+        elif isinstance(self.temperature, (float, int)):
+            self.temperature_fenics = F.as_fenics_constant(
+                self.temperature, self.mesh.mesh
+            )
+        # if temperature is a fem.Constant or function, pass it to temperature_fenics
+        elif isinstance(self.temperature, (fem.Constant, fem.Function)):
+            self.temperature_fenics = self.temperature
+
+        # if temperature is callable, process accordingly
+        elif callable(self.temperature):
+            arguments = self.temperature.__code__.co_varnames
+            if "t" in arguments and "x" not in arguments and "T" not in arguments:
+                # only t is an argument
+                self.temperature_fenics = F.as_fenics_constant(
+                    mesh=self.mesh.mesh, value=self.temperature(t=float(self.t))
+                )
+            else:
+                x = ufl.SpatialCoordinate(self.mesh.mesh)
+                CG_temperature = basix.ufl.element(
+                    basix.ElementFamily.P,
+                    self.mesh.mesh.basix_cell(),
+                    1,
+                    basix.LagrangeVariant.equispaced,
+                )
+                function_space_temperature = fem.FunctionSpace(
+                    self.mesh.mesh, CG_temperature
+                )
+                self.temperature_fenics = fem.Function(function_space_temperature)
+                kwargs = {}
+                if "t" in arguments:
+                    kwargs["t"] = self.t
+                if "x" in arguments:
+                    kwargs["x"] = x
+
+                # store the expression of the temperature
+                # to update the temperature_fenics later
+                self.temperature_expr = fem.Expression(
+                    self.temperature(**kwargs),
+                    function_space_temperature.element.interpolation_points(),
```

Comment:
Do you have to create a functionspace or can you simply use `CG_temperature?`

Reply:
It seems we need to have a functionspace for this one